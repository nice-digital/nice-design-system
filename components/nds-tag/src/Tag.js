import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";

import "./../scss/tag.scss";

export const Tag = props => {
	const {
		children,
		alpha,
		beta,
		live,
		isNew,
		updated,
		consultation,
		impact,
		flush,
		outline
	} = props;

	const classNames = classnames({
		tag: true,
		"tag--alpha": alpha,
		"tag--beta": beta,
		"tag--live": live,
		"tag--new": isNew,
		"tag--updated": updated,
		"tag--consultation": consultation,
		"tag--impact": impact,
		"tag--flush": flush,
		"tag--outline": outline
	});

	return <span className={classNames}>{children}</span>;
};

Tag.propTypes = {
	alpha: PropTypes.bool,
	beta: PropTypes.bool,
	live: PropTypes.bool,
	isNew: PropTypes.bool,
	updated: PropTypes.bool,
	consultation: PropTypes.bool,
	impact: PropTypes.bool,
	flush: PropTypes.bool,
	outline: PropTypes.bool,
	children: PropTypes.oneOfType([
		PropTypes.arrayOf(PropTypes.node),
		PropTypes.node
	]).isRequired
};
