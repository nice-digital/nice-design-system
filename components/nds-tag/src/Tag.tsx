import React, { cloneElement } from "react";
import PropTypes from "prop-types";
import classnames from "classnames";

import "./../scss/tag.scss";

interface BaseTagProps {
	/** Allow any additional props to be passed and applied to the tag */
	[prop: string]: unknown;
	/** Content of the tag */
	children: React.ReactNode;
	/**	Tags have a small margin by default. Use the .tag--flush modifier to remove the margin */
	flush?: boolean;
	/** Use impact tags sparingly, e.g. within phase banners */
	impact?: boolean;
	/** Use the .tag--outline modifier for a tag with a border. Use outline tags with filtered lists to show the currently applied filters. */
	outline?: boolean;
	/** Use removeable tags when users need to be able to remove or dismiss a tag. Use with outline tags at the top of filtered lists to show the currently applied selections. */
	remove?: React.ReactElement;
}

interface UpdatedColour {
	/** Use tags to denote the status of guidance in lists etc */
	updated: boolean;
}

interface AlphaColour {
	/** Use tags to denote service phase */
	alpha: boolean;
}

interface BetaColour {
	/** Use tags to denote service phase */
	beta: boolean;
}

interface ConsultationColour {
	/** Use tags to denote the status of guidance in lists etc */
	consultation: boolean;
}

interface IsNewColour {
	/** Use tags to denote the status of guidance in lists etc */
	isNew: boolean;
}

interface InfoColour {
	/** Use tags to denote status of information */
	info: boolean;
}

interface ErrorColour {
	/** Use tags to denote status of error */
	error: boolean;
}

interface SuccessColour {
	/** Use tags to denote status of success */
	success: boolean;
}
interface CautionColour {
	/** Use tags to denote status of caution */
	caution: boolean;
}

export type TagProps = BaseTagProps &
	(
		| {}
		| UpdatedColour
		| AlphaColour
		| BetaColour
		| ConsultationColour
		| IsNewColour
		| InfoColour
		| ErrorColour
		| SuccessColour
		| CautionColour
	);

export const Tag: React.FC<TagProps> = (props: TagProps) => {
	const {
		children,
		alpha,
		beta,
		isNew,
		updated,
		consultation,
		impact,
		flush,
		outline,
		info,
		error,
		success,
		caution,
		remove,
		...rest
	} = props;

	const classNames = classnames({
		tag: true,
		"tag--alpha": alpha,
		"tag--beta": beta,
		"tag--new": isNew,
		"tag--updated": updated,
		"tag--consultation": consultation,
		"tag--impact": impact,
		"tag--flush": flush,
		"tag--outline": outline,
		"tag--info": info,
		"tag--error": error,
		"tag--success": success,
		"tag--caution": caution
	});

	return (
		<span className={classNames} data-component="tag" {...rest}>
			{children}
			{remove
				? cloneElement(
						remove,
						{ className: classnames(remove.props.className, "tag__remove") },
						<span className="visually-hidden">{remove.props.children}</span>
				  )
				: null}
		</span>
	);
};
