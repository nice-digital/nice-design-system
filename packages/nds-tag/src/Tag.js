import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";

import "./../scss/tag.scss";

export const Tag = props => {
	const {
		children,
		alpha,
		beta,
		live,
		isNew,
		updated,
		consultation,
		impact,
		flush,
		outline,
		info,
		error,
		success,
		caution,
		...rest
	} = props;

	const classNames = classnames({
		tag: true,
		"tag--alpha": alpha,
		"tag--beta": beta,
		"tag--live": live,
		"tag--new": isNew,
		"tag--updated": updated,
		"tag--consultation": consultation,
		"tag--impact": impact,
		"tag--flush": flush,
		"tag--outline": outline,
		"tag--info": info,
		"tag--error": error,
		"tag--success": success,
		"tag--caution": caution
	});

	return (
		<span className={classNames} {...rest}>
			{children}
		</span>
	);
};

Tag.propTypes = {
	alpha: PropTypes.bool,
	beta: PropTypes.bool,
	live: PropTypes.bool,
	isNew: PropTypes.bool,
	updated: PropTypes.bool,
	consultation: PropTypes.bool,
	impact: PropTypes.bool,
	flush: PropTypes.bool,
	outline: PropTypes.bool,
	info: PropTypes.bool,
	error: PropTypes.bool,
	success: PropTypes.bool,
	caution: PropTypes.bool,
	children: PropTypes.oneOfType([
		PropTypes.arrayOf(PropTypes.node),
		PropTypes.node
	]).isRequired
};

Tag.variants = {
	alpha: "alpha",
	beta: "beta",
	live: "live",
	isNew: "isNew",
	updated: "updated",
	consultation: "consultation",
	impact: "impact",
	flush: "flush",
	outline: "outline",
	info: "info",
	error: "error",
	success: "success",
	caution: "caution"
};
