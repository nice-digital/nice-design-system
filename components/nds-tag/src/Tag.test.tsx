import React from "react";
import { render } from "@testing-library/react";

import { Tag } from "./Tag";

describe("Tag", () => {
	it("should match snapshot for default tag", () => {
		const wrapper = render(<Tag>tag content 2</Tag>);
		expect(wrapper).toMatchSnapshot();
	});

	it("should match snapshot for a supplied type", () => {
		const wrapper = render(<Tag alpha>tag content 3</Tag>);
		expect(wrapper).toMatchSnapshot();
	});

	it("should cascade any additional attributes to the container", () => {
		const { container } = render(
			<Tag className="baboon" data-hidden={false}>
				tag content 4
			</Tag>
		);
		expect(
			container.querySelector(".baboon")?.getAttribute("data-hidden")
		).toBe("false");
	});
});
