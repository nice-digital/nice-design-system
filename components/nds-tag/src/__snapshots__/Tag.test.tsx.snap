// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag should match snapshot for a supplied type 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="tag tag--alpha"
        data-component="tag"
      >
        tag content 3
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="tag tag--alpha"
      data-component="tag"
    >
      tag content 3
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Tag should match snapshot for default tag 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="tag"
        data-component="tag"
      >
        tag content 2
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="tag"
      data-component="tag"
    >
      tag content 2
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
